packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,584 @@
+///
+///   final List<String> _kOptions = <String>[
+///     'aardvark',
+///     'bobcat',
+///     'chameleon',
+///   ];
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return AutocompleteCore<String>(
+///       optionsBuilder: (TextEditingValue textEditingValue) {
+///         return _kOptions.where((String option) {
+///           return option.contains(textEditingValue.text.toLowerCase());
+///         });
+///       },
+///       fieldViewBuilder: (BuildContext context, TextEditingController textEditingController, VoidCallback onFieldSubmitted) {
+///         return TextFormField(
+///           controller: textEditingController,
+///           onFieldSubmitted: (String value) {
+///             onFieldSubmitted();
+///           },
+///         );
+///       },
+///       optionsViewBuilder: (BuildContext context, AutocompleteOnSelected<String> onSelected, Iterable<String> options) {
+///         return Material(
+///           elevation: 4.0,
+///           child: Container(
+///             height: 200.0,
+///             child: ListView.builder(
+///               padding: EdgeInsets.all(8.0),
+///               itemCount: options.length,
+///               itemBuilder: (BuildContext context, int index) {
+///                 final String option = options.elementAt(index);
+///                 return GestureDetector(
+///                   onTap: () {
+///                     onSelected(option);
+///                   },
+///                   child: ListTile(
+///                     title: Text(option),
+///                   ),
+///                 );
+///               },
+///             ),
+///           ),
+///         );
+///       },
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+///
+/// {@tool dartpad --template=freeform}
+/// This example is similar to the previous example, but it uses a custom T data
+/// type instead of directly using String.
+///
+/// ```dart imports
+/// import 'package:flutter/widgets.dart';
+/// import 'package:flutter/material.dart';
+/// ```
+///
+/// ```dart
+/// // An example of a type that someone might want to autocomplete a list of.
+/// class User {
+///   const User({
+///     this.email,
+///     this.name,
+///   });
+///
+///   final String email;
+///   final String name;
+///
+///   // The displayStringForOption parameter can be used to get even more
+///   // control over the strings that represent the options objects.
+///   @override
+///   String toString() {
+///     return '$name, $email';
+///   }
+/// }
+///
+/// class AutocompleteCustomTypeExample extends StatelessWidget {
+///   AutocompleteCustomTypeExample({Key key});
+///
+///   final List<User> _kUserOptions = <User>[
+///     User(name: 'Alice', email: 'alice@example.com'),
+///     User(name: 'Bob', email: 'bob@example.com'),
+///     User(name: 'Charlie', email: 'charlie123@gmail.com'),
+///   ];
+///
+///   static String _displayStringForOption(User option) => option.name;
```

Comment:
It looks like the other callbacks could have been defined with static functions if _kUserOptions was static (because why note). Might be best to use one approach or the other.  If you used all static functions then the optionsBuilder could use _displayStringForOption.

Reply:
Sounds good, I'll make it static.

In this example I don't want to use _displayStringForOption in optionsBuilder because I want to show how to search based on email and name while only displaying the name.  I'll add a comment to make that clear.